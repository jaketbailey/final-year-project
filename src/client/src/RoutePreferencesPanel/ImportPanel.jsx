import { useEffect, useState } from "react";

const ImportPanel = (props) => {

    const [file, setFile] = useState(null);
    const [fileType, setFileType] = useState(null);
  

    useEffect(() => {
        // console.log(props.showPanel)
        const panel = document.querySelector('.import-panel');
        if (panel) {
            if (props.showPanel) {
                panel.style.width = 'auto';
                panel.style.height = 'auto';
              } else {
                panel.style.width = '0';
                panel.style.height = '0';
              }
        }
    }, [props.showPanel])

    const extractWaypoints = (data) => {
        const wptArr = []
        const latLngs = [];
    
        if (fileType === 'gpx') {
          // Parse GPX data and extract latLngs of waypoints
          const wptArr = [];
          const parser = new DOMParser();
          const xmlDoc = parser.parseFromString(data, 'text/xml');
          const waypoints = xmlDoc.querySelectorAll('wpt');
      
          waypoints.forEach((waypoint) => {
            const lat = parseFloat(waypoint.getAttribute('lat'));
            const lon = parseFloat(waypoint.getAttribute('lon'));
            latLngs.push(L.latLng(lat, lon));
            wptArr.push([lat, lon]);
          });
    
        } else if (fileType === 'geojson') {
          const geoJSON = JSON.parse(data);
          geoJSON.features.forEach((feature) => {
              const type = feature.geometry.type;
              const coordinates = feature.geometry.coordinates;
              if (type === 'Point') {
              latLngs.push(L.latLng(coordinates[1], coordinates[0]));
                wptArr.push([coordinates[1], coordinates[0]]);
              }
          });
        }
        localStorage.setItem('waypoints', wptArr)
        props.control.current.setWaypoints(latLngs);
      };
    
    
      const handleFileChange = (e) => {
        console.log(e.target.files)
        const file = e.target.files[0];
        const fileDetails = e.target.value.split('.')
        const type = fileDetails[fileDetails.length - 1];
        setFile(file);
        setFileType(type);
      }
    
      const handleFileUpload = (e) => {
        if (file !== null) {
          const reader = new FileReader();
          reader.onload = (e) => {
            if (fileType === 'geojson') {
              const currentFile = e.target.result;
              extractWaypoints(currentFile);
              return;
            } else if (fileType === 'gpx') {
              const currentFile = e.target.result;
              extractWaypoints(currentFile);
              return;
            }
          };
          
          reader.readAsText(file);
        }
      }

    return (
        <div className='import-panel'>
            <input className='importFileSelect' type='file' onChange={handleFileChange} />
            <button className='importFileBtn' onClick={handleFileUpload}>Load File</button>
        </div>
    )
}

export default ImportPanel;