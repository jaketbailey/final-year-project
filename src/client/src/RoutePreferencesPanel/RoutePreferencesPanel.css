

.route-preferences-panel, .share-panel {
    position: absolute; 
    background-color: white;
    border: 1px solid #ccc;
    padding: 0;
    width: fit-content;
    height: fit-content;
    left: 50%;
    transform: translate(-50%, 0);
}

.route-preferences-panel__preferences, .share-panel__preferences {
    overflow: hidden;
    font: 0.8rem Poppins;
}

  .checkbox-list, .button-list {
    display: flex;
    flex-direction: column;
  }

  .checkbox-item {
    display: flex;
    justify-content: left;
    align-items: center;
    margin-bottom: 5px;
    margin-left: 1rem;
  }

  .checkbox-label {
    margin-left: 10px;
  }

p {
    font: 0.8rem Poppins;
    padding: 0.5rem;
    margin: 0.5rem;
}

.left {
    float: left;
    width: 15%;
    text-align: right;
}

.right {
    float: right;
    width: 85%;
    text-align: left;
}

.route-preferences-panel__button, .share-panel__button {
    all: none;
    font: 0.8rem Poppins;
    background-color: white;
    border: 0px;
    cursor: pointer;
    margin: 0.2rem;
    /* writing-mode: vertical-lr;
    text-orientation: upright; */
}

.route-preferences-panel__button:hover, .share-panel__button:hover {
    background-color: #ccc;
    cursor: pointer;
}

.route-preferences-panel__button, .share-panel__button {
    height: auto;
}

