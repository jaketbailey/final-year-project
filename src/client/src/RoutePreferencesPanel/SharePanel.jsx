import { useRef, useState } from 'react';
import './RoutePreferencesPanel.css'
import { useEffect } from 'react';
import Modal from '../Modal/Modal';

const SharePanel = (props) => {
  useEffect(() => {
    const sharePanelContainer = document.querySelector('.share-panel__preferences');
    if (props.showPanel) {
      sharePanelContainer.style.width = 'auto';
      sharePanelContainer.style.height = 'auto';
    } else {
      sharePanelContainer.style.width = '0';
      sharePanelContainer.style.height = '0';
    }
  }, [props.showPanel]);

  const message = useRef(null);
  const routeGeoJSON = useRef(props.geoJSON);
  const routeGPX = useRef(props.gpx);

  useEffect(() => {
    console.log('update to route')
    routeGeoJSON.current = props.geoJSON;
    routeGPX.current = props.gpx;
    console.log(routeGPX)
    console.log(routeGeoJSON)

    message.current = {
      to: '',
      from: 'jake@jaketbailey.co.uk',
      subject: 'Route',
      text: 'Planned Route using the UP2002753 Route Planner',
      attachments: [
        {
          content: routeGeoJSON.current,
          filename: '',
          type: 'application/geojson',
          disposition: 'attachment'
        },
        {
          content: routeGPX.current,
          filename: '',
          type: 'application/gpx+xml',
          disposition: 'attachment'
        }
      ]
    }
    console.log(message)
  }, [props.geoJSON, props.gpx]);

  useEffect(() => {
    message.current.to = props.data.to;
    message.current.attachments[0].filename = `${props.data.filename}.geojson`;
    message.current.attachments[1].filename = `${props.data.filename}.gpx`;
    console.log('hello message')
    console.log(message.current)
    sendEmail();
  },[props.data]);

  useEffect(() => {
    const shareEmailButton = document.querySelector('#shareEmailButton');
    shareEmailButton.addEventListener('click', clickHandler);
  }, []);

  const sendEmail = async () => {
    if (message.current === null) {
      return;
    }

    const body = JSON.stringify(message.current);
    console.log(body);
    const response = await fetch('/api/send-email', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: body,
    });
    const res = await response.json();
    console.log(res);
  }

  const hideEmailModal = () => {
    const modal = document.querySelector('.modal');
    modal.style.display = 'none';
  }
  
  // const [show, setShow] = useState(false);

  const clickHandler = (event) => {
    const id = event.target.id;
    if (id === 'shareEmailButton') {
      console.log('email');
      props.setShow(!props.show);
    };
  };

  return (
    <div className="share-panel__preferences">
      <div className="share-panel__preferences__preference">
        <div className='checkbox-list'>
          <button id='shareEmailButton' type='button' className='share-panel__button'>Email</button>
        </div>
      </div>
    </div>
  )
}

export default SharePanel;