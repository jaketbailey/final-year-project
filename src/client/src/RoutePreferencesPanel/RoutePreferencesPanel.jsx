import { useEffect, useState } from 'react';
import './RoutePreferencesPanel.css'
import 'leaflet-routing-machine';
import SharePanel from './SharePanel';

/**
 * @function RoutePreferencesPanel
 * @description Creates a panel for the user to select route preferences
 * @param {*} props
 * @returns RoutePreferencesPanel component
 */
const RoutePreferencesPanel = (props) => {
  const [showPanel, setShowPanel] = useState(false);
  const [showSharePanel, setShowSharePanel] = useState(false);
  const [avoidFeatures, setAvoidFeatures] = useState([]);
  const [stravaAuthCode, setStravaAuthCode] = useState(null)
 
  const deauthoriseStrava = async () => {
    const ACCESS_TOKEN = props.stravaAccessToken.access_token;
    const URL = `https://www.strava.com/oauth/deauthorize?access_token=${ACCESS_TOKEN}`
    const response = await fetch(URL, {
      method: 'POST',
    })
    const res = await response.json();;
    console.log(res)
    localStorage.removeItem('strava_auth_code');
    localStorage.removeItem('strava_access_token');
    setStravaAuthCode(null);
    props.setStravaAccessToken(null);
  }

  const getStravaAuthCode = () => {
    const stravaBtn = document.querySelector('#strava-login');
    if (stravaBtn.classList.contains('strava-logout')) {
      deauthoriseStrava();
      stravaBtn.textContent = 'Log in to Strava'
      stravaBtn.classList.remove('strava-logout');
      return;
    }
    const CLIENT_ID = import.meta.env.VITE_STRAVA_CLIENT_ID;
    const REDIRECT_URI = import.meta.env.VITE_STRAVA_REDIRECT_URI;
    const SCOPE = 'activity:read_all,activity:write';
    // Redirect to get auth code
    const URL = `https://www.strava.com/oauth/authorize?client_id=${CLIENT_ID}&redirect_uri=${REDIRECT_URI}&response_type=code&scope=${SCOPE}`
    window.location.href = URL;
  }

  useEffect(() => {
    if (stravaAuthCode === null) {
      localStorage.removeItem('strava_auth_code');
      return;
    }
    getStravaAccessToken();
  }, [stravaAuthCode]);

  const getStravaAccessToken = async () => {
    const CLIENT_ID = import.meta.env.VITE_STRAVA_CLIENT_ID;
    const CLIENT_SECRET = import.meta.env.VITE_STRAVA_CLIENT_SECRET;
    const CODE = stravaAuthCode.code;
    const GRANT_TYPE = 'authorization_code';
    const URL = `https://www.strava.com/api/v3/oauth/token?client_id=${CLIENT_ID}&client_secret=${CLIENT_SECRET}&code=${CODE}&grant_type=${GRANT_TYPE}`
    
    const response = await fetch(URL, {
      method: 'POST',
    });
    const res = await response.json();
    if (res.message === 'Bad Request') {
      getStravaAuthCode();
      return;
    }
    props.setStravaAccessToken(res);
    localStorage.setItem('strava_access_token', JSON.stringify(res));
    return;
  } 

  useEffect(() => {
    const checkboxes = document.querySelectorAll('.checkbox-input');
    for (const checkbox of checkboxes) {
      checkbox.addEventListener('change', (event) => {
        if (event.target.checked) {
          setAvoidFeatures(avoidFeatures => [...avoidFeatures, event.target.value]);
        } else {
          setAvoidFeatures(avoidFeatures => avoidFeatures.filter(feature => feature !== event.target.value));
        }
      });
    }

    const stravaAccessToken = JSON.parse(localStorage.getItem('strava_access_token'));
    if (stravaAccessToken) {
      props.setStravaAccessToken(stravaAccessToken);
      const stravaBtn = document.querySelector("#strava-login");
      if (stravaBtn !== null) {
        stravaBtn.innerText = 'Log out of Strava';
        stravaBtn.classList.add('strava-logout');
      }
      return;
    }

    const pathname = window.location.pathname;
    if (pathname === '/exchange_token') {
      const params = (new URL(document.location)).searchParams;
      console.log(params)
      localStorage.setItem('strava_auth_code', params.get('code'))
      if (params.get('error')) {
        console.log(params.get('error'))
        return;
      }
      const stravaBtn = document.querySelector("#strava-login");
      if (stravaBtn !== null) {
        stravaBtn.innerText = 'Log out of Strava';
        stravaBtn.classList.add('strava-logout');
      }
      const code = {
         code: params.get('code'),
         scope: params.get('scope'),
      }

      localStorage.setItem('strava_auth_code', JSON.stringify(code));
      setStravaAuthCode(code);
      return;
    }
    return;
  }, []);

  useEffect(() => {
    if (props.control.current === null) {
      return;
    }
    const filterAvoidFeatures = [...new Set(avoidFeatures)];
    props.control.current.getRouter().options.routingQueryParams.options.avoid_features = filterAvoidFeatures;
    props.control.current.route();
  }, [avoidFeatures]);

  const togglePanel = () => {
    setShowPanel(!showPanel);
  }

  const toggleSharePanel = () => {
    setShowSharePanel(!showSharePanel);
  }

  const saveGeoJSON = () => {
    props.geoJSONLink.click();
  }

  const saveGPX = () => {
      props.gpxLink.click();
  }

  useEffect(() => {
    const routePreferencesPanelContainer = document.querySelector('.route-preferences-panel__preferences');
    if (showPanel) {
      routePreferencesPanelContainer.style.width = 'auto';
      routePreferencesPanelContainer.style.height = 'auto';
    } else {
      routePreferencesPanelContainer.style.width = '0';
      routePreferencesPanelContainer.style.height = '0';
    }
  }, [showPanel]);

  return (
      <div className="route-preferences-panel">
        <button className="route-preferences-panel__button" onClick={togglePanel}>
          Route Preferences 
        </button>
        <button className="route-preferences-panel__button" onClick={saveGeoJSON} >
          Export Route as GeoJSON
        </button>
        <button className="route-preferences-panel__button" onClick={saveGPX} >
          Export Route as GPX
        </button> 
        <button id="strava-login" className="route-preferences-panel__button" onClick={getStravaAuthCode}>
          Log in to Strava 
        </button>
        <button className="route-preferences-panel__button" onClick={toggleSharePanel} >
          Share
        </button>
        <div className="route-preferences-panel__preferences">
          <div className="route-preferences-panel__preferences__preference">
            <p>Avoid:</p>
            <div className='checkbox-list'>
              <div className='checkbox-item'>
                <input type='checkbox' className="checkbox-input" id='avoidSteps' name='Steps' value='steps' />
                <label className='checkbox-label' htmlFor='avoidSteps'>Steps</label>
              </div>
              <div className='checkbox-item'>
                <input type='checkbox' className="checkbox-input"  id='avoidFerries' name='Ferries' value='ferries' />
                <label className='checkbox-label' htmlFor='avoidFerries'>Ferries</label>
              </div>
              <div className='checkbox-item'>
                <input type='checkbox' className="checkbox-input"  id='avoidFords' name='Fords' value='fords' />
                <label className='checkbox-label' htmlFor='avoidFords'>Fords</label>
              </div>
            </div>
          </div>
        </div>
        <SharePanel 
          geoJSON={props.geoJSON} 
          gpx={props.gpx} 
          showPanel={showSharePanel}
          setShow={props.setShow}
          show={props.show}
          setShowStrava={props.setShowStrava}
          showStrava={props.showStrava}
          data={props.emailData}
          stravaAccessToken={props.stravaAccessToken}
          stravaData={props.stravaData}
        />
      </div>
  );
}

export default RoutePreferencesPanel;