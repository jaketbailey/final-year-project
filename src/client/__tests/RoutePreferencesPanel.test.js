import RoutePreferencesPanel from "../src/RoutePreferencesPanel/RoutePreferencesPanel";
import { render, act } from "@testing-library/react";

describe("RoutePreferencesPanel", () => {
    it("should render", () => {
        const geoJSONLink = "";
        const gpxLink = "";
        const setAvoidFeatures = jest.fn();
        const avoidFeatures = [];
        const control = {current: {
            getRouter: jest.fn(() => ({
                getFeatures: jest.fn(() => []),
                options: {
                    routingQueryParams: {
                        options: {
                            avoid_features: []
                        },
                    },
                }
            })),
            route: jest.fn(),
        }};


        const component = render(
            <RoutePreferencesPanel
                geoJSONLink={geoJSONLink} 
                gpxLink={gpxLink}
                setAvoidFeatures={setAvoidFeatures}
                avoidFeatures={avoidFeatures}
                control={control}
            />
        );

        expect(component).toMatchSnapshot();
    });
});