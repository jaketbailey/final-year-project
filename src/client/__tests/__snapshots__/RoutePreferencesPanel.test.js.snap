// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoutePreferencesPanel should render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="route-preferences-panel"
      >
        <button
          class="route-preferences-panel__button"
        >
          Route Preferences
        </button>
        <button
          class="route-preferences-panel__button"
        >
          Export Route as GeoJSON
        </button>
        <button
          class="route-preferences-panel__button"
        >
          Export Route as GPX
        </button>
        <button
          class="route-preferences-panel__button"
        >
          Log in to Strava
        </button>
        <button
          class="route-preferences-panel__button"
        >
          Share
        </button>
        <div
          class="route-preferences-panel__preferences"
          style="width: 0px; height: 0px;"
        >
          <div
            class="route-preferences-panel__preferences__preference"
          >
            <p>
              Avoid:
            </p>
            <div
              class="checkbox-list"
            >
              <div
                class="checkbox-item"
              >
                <input
                  class="checkbox-input"
                  id="avoidSteps"
                  name="Steps"
                  type="checkbox"
                  value="steps"
                />
                <label
                  class="checkbox-label"
                  for="avoidSteps"
                >
                  Steps
                </label>
              </div>
              <div
                class="checkbox-item"
              >
                <input
                  class="checkbox-input"
                  id="avoidFerries"
                  name="Ferries"
                  type="checkbox"
                  value="ferries"
                />
                <label
                  class="checkbox-label"
                  for="avoidFerries"
                >
                  Ferries
                </label>
              </div>
              <div
                class="checkbox-item"
              >
                <input
                  class="checkbox-input"
                  id="avoidFords"
                  name="Fords"
                  type="checkbox"
                  value="fords"
                />
                <label
                  class="checkbox-label"
                  for="avoidFords"
                >
                  Fords
                </label>
              </div>
            </div>
          </div>
        </div>
        <div
          class="share-panel__preferences"
          style="width: 0px; height: 0px;"
        >
          <div
            class="share-panel__preferences__preference"
          >
            <div
              class="button-list"
            >
              <button
                class="share-panel__button"
                id="shareEmailButton"
                type="button"
              >
                Email
              </button>
              <button
                class="share-panel__button"
                id="shareStravaButton"
                type="button"
              >
                Strava Activity
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="route-preferences-panel"
    >
      <button
        class="route-preferences-panel__button"
      >
        Route Preferences
      </button>
      <button
        class="route-preferences-panel__button"
      >
        Export Route as GeoJSON
      </button>
      <button
        class="route-preferences-panel__button"
      >
        Export Route as GPX
      </button>
      <button
        class="route-preferences-panel__button"
      >
        Log in to Strava
      </button>
      <button
        class="route-preferences-panel__button"
      >
        Share
      </button>
      <div
        class="route-preferences-panel__preferences"
        style="width: 0px; height: 0px;"
      >
        <div
          class="route-preferences-panel__preferences__preference"
        >
          <p>
            Avoid:
          </p>
          <div
            class="checkbox-list"
          >
            <div
              class="checkbox-item"
            >
              <input
                class="checkbox-input"
                id="avoidSteps"
                name="Steps"
                type="checkbox"
                value="steps"
              />
              <label
                class="checkbox-label"
                for="avoidSteps"
              >
                Steps
              </label>
            </div>
            <div
              class="checkbox-item"
            >
              <input
                class="checkbox-input"
                id="avoidFerries"
                name="Ferries"
                type="checkbox"
                value="ferries"
              />
              <label
                class="checkbox-label"
                for="avoidFerries"
              >
                Ferries
              </label>
            </div>
            <div
              class="checkbox-item"
            >
              <input
                class="checkbox-input"
                id="avoidFords"
                name="Fords"
                type="checkbox"
                value="fords"
              />
              <label
                class="checkbox-label"
                for="avoidFords"
              >
                Fords
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="share-panel__preferences"
        style="width: 0px; height: 0px;"
      >
        <div
          class="share-panel__preferences__preference"
        >
          <div
            class="button-list"
          >
            <button
              class="share-panel__button"
              id="shareEmailButton"
              type="button"
            >
              Email
            </button>
            <button
              class="share-panel__button"
              id="shareStravaButton"
              type="button"
            >
              Strava Activity
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
